File: cpp/ems-core/test/journal_test.cpp
// Journal codec tests.
//
// Every expectation here is deliberately the same assertion as
// java/ems-core/.../JournalCodecTest.java and rust/ems-core/tests/journal_test.rs.
// Three implementations agreeing on their own unit tests is what makes the
// byte-exact conformance run a formality rather than a surprise.

#include "ems_core/journal.hpp"

#include <gtest/gtest.h>

#include <filesystem>
#include <fstream>
#include <sstream>
#include <string>
#include <vector>

// POSIX getpid(). The CI runners and the devcontainer are both Linux; if this
// tree ever targets Windows, swap it for std::this_thread::get_id().
#include <unistd.h>

namespace {

using ems::core::decode;
using ems::core::encode;
using ems::core::JournalEvent;
using ems::core::read_journal;
using ems::core::write_journal;

std::filesystem::path temp_file(const std::string& name) {
    static int counter = 0;
    const auto dir = std::filesystem::temp_directory_path() /
                     ("ems-core-journal-test-" + std::to_string(::getpid()) + "-" +
                      std::to_string(++counter));
    // Process id as well as a counter: a fixed name is reused across runs, so
    // a previous run's output file was still present and a test asserting
    // "this file must not exist yet" failed on a file it did not create.
    std::filesystem::remove_all(dir);
    std::filesystem::create_directories(dir);
    return dir / name;
}

std::string read_file(const std::filesystem::path& path) {
    std::ifstream in(path, std::ios::binary);
    std::ostringstream buffer;
    buffer << in.rdbuf();
    return buffer.str();
}

JournalEvent event(std::uint64_t seq, const std::string& type,
                   std::map<std::string, std::string> fields = {}) {
    return JournalEvent{seq, type, std::move(fields)};
}

TEST(Journal, WritesTopLevelAndFieldKeysInLexicographicOrder) {
    const auto path = temp_file("out.jsonl");
    ASSERT_TRUE(write_journal(path, {event(1, "OrderNew", {{"qty", "100"}, {"account", "ACC1"}})}));

    EXPECT_EQ(read_file(path),
              "{\"fields\":{\"account\":\"ACC1\",\"qty\":\"100\"},\"seq\":1,\"type\":\"OrderNew\"}\n");
}

TEST(Journal, EveryLineEndsWithNewlineIncludingTheLast) {
    const auto path = temp_file("out.jsonl");
    ASSERT_TRUE(write_journal(path, {event(1, "A"), event(2, "B")}));

    EXPECT_EQ(read_file(path),
              "{\"fields\":{},\"seq\":1,\"type\":\"A\"}\n{\"fields\":{},\"seq\":2,\"type\":\"B\"}\n");
}

TEST(Journal, RoundTripsWithoutLoss) {
    const auto path = temp_file("rt.jsonl");
    const std::vector<JournalEvent> events = {
        event(1, "OrderNew", {{"figi", "BBG000B9XRY4"}, {"price", "1250000"}}),
        event(2, "OrderAccepted", {{"orderId", "ORD-0000000001"}})};
    ASSERT_TRUE(write_journal(path, events));

    const auto read = read_journal(path);
    ASSERT_TRUE(read.has_value()) << read.error().to_string();
    EXPECT_EQ(read.value(), events);
}

TEST(Journal, EscapesQuotesBackslashesAndControlCharacters) {
    const auto path = temp_file("esc.jsonl");
    const std::string raw = std::string("a\"b\\c\nd\te") + '\x01' + "f";
    ASSERT_TRUE(write_journal(path, {event(1, "Note", {{"text", raw}})}));

    EXPECT_EQ(read_file(path),
              "{\"fields\":{\"text\":\"a\\\"b\\\\c\\nd\\te\\u0001f\"},\"seq\":1,\"type\":\"Note\"}\n");

    const auto read = read_journal(path);
    ASSERT_TRUE(read.has_value()) << read.error().to_string();
    EXPECT_EQ(read.value().at(0).fields.at("text"), raw);
}

TEST(Journal, NonAsciiIsWrittenAsUtf8NotEscaped) {
    const auto path = temp_file("utf8.jsonl");
    ASSERT_TRUE(write_journal(path, {event(1, "Note", {{"text", "café — ☕"}})}));

    EXPECT_EQ(read_file(path), "{\"fields\":{\"text\":\"café — ☕\"},\"seq\":1,\"type\":\"Note\"}\n");

    const auto read = read_journal(path);
    ASSERT_TRUE(read.has_value()) << read.error().to_string();
    EXPECT_EQ(read.value().at(0).fields.at("text"), "café — ☕");
}

TEST(Journal, UnicodeEscapeDecodesToUtf8) {
    // U+2615 HOT BEVERAGE, written by another producer as an escape. Decoding
    // it must yield the same bytes we would have written directly.
    const auto parsed = decode("{\"fields\":{\"t\":\"\\u2615\"},\"seq\":1,\"type\":\"A\"}", 1);
    ASSERT_TRUE(parsed.has_value()) << parsed.error().to_string();
    EXPECT_EQ(parsed.value().fields.at("t"), "☕");
}

TEST(Journal, EmptyJournalRoundTrips) {
    const auto path = temp_file("empty.jsonl");
    ASSERT_TRUE(write_journal(path, {}));

    EXPECT_EQ(read_file(path), "");
    const auto read = read_journal(path);
    ASSERT_TRUE(read.has_value());
    EXPECT_TRUE(read.value().empty());
}

TEST(Journal, BlankLinesAreIgnoredOnRead) {
    const auto path = temp_file("blank.jsonl");
    {
        std::ofstream out(path, std::ios::binary);
        out << "{\"fields\":{},\"seq\":1,\"type\":\"A\"}\n\n";
    }

    const auto read = read_journal(path);
    ASSERT_TRUE(read.has_value()) << read.error().to_string();
    EXPECT_EQ(read.value().size(), 1U);
}

TEST(Journal, MalformedLineReportsItsLineNumber) {
    const auto path = temp_file("bad.jsonl");
    {
        std::ofstream out(path, std::ios::binary);
        out << "{\"fields\":{},\"seq\":1,\"type\":\"A\"}\nnot json\n";
    }

    const auto read = read_journal(path);
    ASSERT_FALSE(read.has_value());
    EXPECT_EQ(read.error().line, 2U);
}

TEST(Journal, UnknownTopLevelKeyIsRejected) {
    const auto parsed = decode("{\"fields\":{},\"seq\":1,\"type\":\"A\",\"extra\":\"x\"}", 1);
    ASSERT_FALSE(parsed.has_value());
    EXPECT_NE(parsed.error().message.find("extra"), std::string::npos);
}

TEST(Journal, MissingRequiredKeyIsRejected) {
    const auto parsed = decode("{\"fields\":{},\"seq\":1}", 1);
    ASSERT_FALSE(parsed.has_value());
    EXPECT_NE(parsed.error().message.find("type"), std::string::npos);
}

TEST(Journal, DuplicateTopLevelKeyIsRejected) {
    const auto parsed = decode("{\"fields\":{},\"seq\":1,\"seq\":2,\"type\":\"A\"}", 1);
    ASSERT_FALSE(parsed.has_value());
    EXPECT_NE(parsed.error().message.find("duplicate"), std::string::npos);
}

TEST(Journal, NonStringFieldValueIsRejected) {
    EXPECT_FALSE(decode("{\"fields\":{\"qty\":100},\"seq\":1,\"type\":\"A\"}", 1).has_value());
}

TEST(Journal, TrailingContentAfterTheObjectIsRejected) {
    EXPECT_FALSE(decode("{\"fields\":{},\"seq\":1,\"type\":\"A\"} junk", 1).has_value());
}

TEST(Journal, NegativeSequenceIsRejected) {
    EXPECT_FALSE(decode("{\"fields\":{},\"seq\":-1,\"type\":\"A\"}", 1).has_value());
}

TEST(Journal, SequenceOverflowIsRejectedRatherThanWrapped) {
    // A wrapped sequence number is a journal that replays in a different order.
    EXPECT_FALSE(
        decode("{\"fields\":{},\"seq\":99999999999999999999999,\"type\":\"A\"}", 1).has_value());
}

TEST(Journal, LoneSurrogateEscapeIsRejected) {
    // Legal JSON, but not a Unicode scalar value — Rust's char cannot hold one,
    // so accepting it here would diverge from the Rust journal.
    const auto parsed = decode("{\"fields\":{},\"seq\":1,\"type\":\"\\ud800\"}", 1);
    ASSERT_FALSE(parsed.has_value());
    EXPECT_NE(parsed.error().message.find("scalar value"), std::string::npos);
}

TEST(Journal, HostileInputNeverCrashes) {
    // The parser is a fuzz target. Every one of these must be a clean rejection,
    // and under ASan/UBSan this test is also the out-of-bounds check.
    const std::vector<std::string> cases = {"",
                                            "{",
                                            "}",
                                            "{\"",
                                            "{\"fields\"",
                                            "{\"fields\":",
                                            "{\"fields\":{",
                                            "{\"fields\":{\"a\"",
                                            "{\"fields\":{\"a\":",
                                            "{\"fields\":{\"a\":\"",
                                            "{\"fields\":{},\"seq\":",
                                            "{\"fields\":{},\"seq\":1,\"type\":\"\\u\"}",
                                            "{\"fields\":{},\"seq\":1,\"type\":\"\\uZZZZ\"}",
                                            "{\"fields\":{},\"seq\":1,\"type\":\"\\q\"}",
                                            "{\"fields\":{},\"seq\":1,\"type\":\"é",
                                            std::string(1, '\0'),
                                            "[]",
                                            "null"};
    for (const auto& input : cases) {
        EXPECT_FALSE(decode(input, 1).has_value()) << "expected rejection for: " << input;
    }
}

TEST(Journal, VeryLongInputIsHandledWithoutCrash) {
    const std::string long_value(100000, 'x');
    const auto parsed =
        decode("{\"fields\":{\"a\":\"" + long_value + "\"},\"seq\":1,\"type\":\"A\"}", 1);
    ASSERT_TRUE(parsed.has_value()) << parsed.error().to_string();
    EXPECT_EQ(parsed.value().fields.at("a").size(), 100000U);
}

TEST(Journal, EncodeMatchesTheAgreedByteLayout) {
    EXPECT_EQ(encode(event(7, "Fill", {{"qty", "50"}})),
              "{\"fields\":{\"qty\":\"50\"},\"seq\":7,\"type\":\"Fill\"}");
}

}  // namespace

File: rust/ems-core/tests/journal_test.rs
//! Journal codec tests.
//!
//! Every expectation here is deliberately the same assertion as
//! `java/ems-core/src/test/java/io/crossasset/ems/core/journal/JournalCodecTest.java`.
//! Two implementations agreeing on their own unit tests is what makes the
//! byte-exact conformance run a formality rather than a surprise.

// A failing test SHOULD panic — that is how it reports. The workspace denies
// expect/unwrap/panic because the slice binary must turn errors into an exit
// code or a journal event, and that rule has no business applying to tests.
#![allow(clippy::expect_used, clippy::unwrap_used, clippy::panic)]

use std::fs;
use std::path::PathBuf;
use std::sync::atomic::{AtomicU64, Ordering};

use ems_core::journal::{decode, read_journal, write_journal, JournalError, JournalEvent};

fn tmp_dir() -> PathBuf {
    static COUNTER: AtomicU64 = AtomicU64::new(0);
    // Process id as well as a counter, and the directory is cleared: reusing a
    // fixed name across runs meant a previous run's out.jsonl was still there,
    // and `publish_before_flush_writes_nothing` failed on a file it did not
    // create. Wall-clock is not an option — the port bans it.
    let dir = std::env::temp_dir().join(format!(
        "ems-{}-{}-{}",
        env!("CARGO_PKG_NAME"),
        std::process::id(),
        COUNTER.fetch_add(1, Ordering::Relaxed)
    ));
    let _ = fs::remove_dir_all(&dir);
    fs::create_dir_all(&dir).expect("create temp dir");
    dir
}

fn event(seq: u64, event_type: &str, fields: &[(&str, &str)]) -> JournalEvent {
    JournalEvent {
        seq,
        event_type: event_type.to_owned(),
        fields: fields
            .iter()
            .map(|(k, v)| ((*k).to_owned(), (*v).to_owned()))
            .collect(),
    }
}

#[test]
fn writes_top_level_and_field_keys_in_lexicographic_order() {
    let dir = tmp_dir();
    let path = dir.join("out.jsonl");
    write_journal(
        &path,
        &[event(1, "OrderNew", &[("qty", "100"), ("account", "ACC1")])],
    )
    .expect("write");

    assert_eq!(
        fs::read_to_string(&path).expect("read"),
        "{\"fields\":{\"account\":\"ACC1\",\"qty\":\"100\"},\"seq\":1,\"type\":\"OrderNew\"}\n"
    );
}

#[test]
fn every_line_ends_with_newline_including_the_last() {
    let dir = tmp_dir();
    let path = dir.join("out.jsonl");
    write_journal(&path, &[event(1, "A", &[]), event(2, "B", &[])]).expect("write");

    assert_eq!(
        fs::read_to_string(&path).expect("read"),
        "{\"fields\":{},\"seq\":1,\"type\":\"A\"}\n{\"fields\":{},\"seq\":2,\"type\":\"B\"}\n"
    );
}

#[test]
fn round_trips_without_loss() {
    let dir = tmp_dir();
    let path = dir.join("rt.jsonl");
    let events = vec![
        event(
            1,
            "OrderNew",
            &[("figi", "BBG000B9XRY4"), ("price", "1250000")],
        ),
        event(2, "OrderAccepted", &[("orderId", "ORD-0000000001")]),
    ];
    write_journal(&path, &events).expect("write");

    assert_eq!(read_journal(&path).expect("read"), events);
}

#[test]
fn escapes_quotes_backslashes_and_control_characters() {
    let dir = tmp_dir();
    let path = dir.join("esc.jsonl");
    let raw = "a\"b\\c\nd\te\u{1}f";
    write_journal(&path, &[event(1, "Note", &[("text", raw)])]).expect("write");

    assert_eq!(
        fs::read_to_string(&path).expect("read"),
        "{\"fields\":{\"text\":\"a\\\"b\\\\c\\nd\\te\\u0001f\"},\"seq\":1,\"type\":\"Note\"}\n"
    );
    assert_eq!(read_journal(&path).expect("read")[0].fields["text"], raw);
}

#[test]
fn non_ascii_is_written_as_utf8_not_escaped() {
    let dir = tmp_dir();
    let path = dir.join("utf8.jsonl");
    write_journal(&path, &[event(1, "Note", &[("text", "café — ☕")])]).expect("write");

    assert_eq!(
        fs::read_to_string(&path).expect("read"),
        "{\"fields\":{\"text\":\"café — ☕\"},\"seq\":1,\"type\":\"Note\"}\n"
    );
    assert_eq!(
        read_journal(&path).expect("read")[0].fields["text"],
        "café — ☕"
    );
}

#[test]
fn empty_journal_round_trips() {
    let dir = tmp_dir();
    let path = dir.join("empty.jsonl");
    write_journal(&path, &[]).expect("write");

    assert_eq!(fs::read_to_string(&path).expect("read"), "");
    assert!(read_journal(&path).expect("read").is_empty());
}

#[test]
fn blank_lines_are_ignored_on_read() {
    let dir = tmp_dir();
    let path = dir.join("blank.jsonl");
    fs::write(&path, "{\"fields\":{},\"seq\":1,\"type\":\"A\"}\n\n").expect("write");

    assert_eq!(read_journal(&path).expect("read").len(), 1);
}

#[test]
fn malformed_line_reports_its_line_number() {
    let dir = tmp_dir();
    let path = dir.join("bad.jsonl");
    fs::write(
        &path,
        "{\"fields\":{},\"seq\":1,\"type\":\"A\"}\nnot json\n",
    )
    .expect("write");

    match read_journal(&path) {
        Err(JournalError::Malformed(e)) => assert_eq!(e.line, 2),
        other => panic!("expected a malformed-journal error, got {other:?}"),
    }
}

#[test]
fn unknown_top_level_key_is_rejected() {
    let err = decode(
        "{\"fields\":{},\"seq\":1,\"type\":\"A\",\"extra\":\"x\"}",
        1,
    )
    .expect_err("must reject");
    assert!(err.message.contains("extra"), "{err}");
}

#[test]
fn missing_required_key_is_rejected() {
    let err = decode("{\"fields\":{},\"seq\":1}", 1).expect_err("must reject");
    assert!(err.message.contains("type"), "{err}");
}

#[test]
fn duplicate_top_level_key_is_rejected() {
    let err =
        decode("{\"fields\":{},\"seq\":1,\"seq\":2,\"type\":\"A\"}", 1).expect_err("must reject");
    assert!(err.message.contains("duplicate"), "{err}");
}

#[test]
fn non_string_field_value_is_rejected() {
    decode("{\"fields\":{\"qty\":100},\"seq\":1,\"type\":\"A\"}", 1).expect_err("must reject");
}

#[test]
fn trailing_content_after_the_object_is_rejected() {
    decode("{\"fields\":{},\"seq\":1,\"type\":\"A\"} junk", 1).expect_err("must reject");
}

#[test]
fn negative_sequence_is_rejected() {
    decode("{\"fields\":{},\"seq\":-1,\"type\":\"A\"}", 1).expect_err("must reject");
}

#[test]
fn fields_are_ordered_regardless_of_insertion_order() {
    let a = event(1, "X", &[("b", "2"), ("a", "1")]);
    let b = event(1, "X", &[("a", "1"), ("b", "2")]);

    assert_eq!(a, b);
    assert_eq!(a.fields.keys().collect::<Vec<_>>(), vec!["a", "b"]);
}

/// Hostile input: the parser is a fuzz target, so none of these may panic.
#[test]
fn hostile_input_never_panics() {
    let cases = [
        "",
        "{",
        "}",
        "{\"",
        "{\"fields\"",
        "{\"fields\":",
        "{\"fields\":{",
        "{\"fields\":{\"a\"",
        "{\"fields\":{\"a\":",
        "{\"fields\":{\"a\":\"",
        "{\"fields\":{},\"seq\":",
        "{\"fields\":{},\"seq\":99999999999999999999999,\"type\":\"A\"}",
        "{\"fields\":{},\"seq\":1,\"type\":\"\\u\"}",
        "{\"fields\":{},\"seq\":1,\"type\":\"\\uZZZZ\"}",
        "{\"fields\":{},\"seq\":1,\"type\":\"\\q\"}",
        "{\"fields\":{},\"seq\":1,\"type\":\"é",
        "\u{0}",
        "[]",
        "null",
    ];
    for case in cases {
        let result = decode(case, 1);
        assert!(result.is_err(), "expected rejection for {case:?}");
    }
}

/// A lone surrogate is a valid `\uXXXX` escape in JSON but not a Rust `char`.
/// Rejecting it is the honest answer; silently substituting U+FFFD would put a
/// different byte sequence in the journal than Java produces.
#[test]
fn lone_surrogate_escape_is_rejected() {
    let err = decode("{\"fields\":{},\"seq\":1,\"type\":\"\\ud800\"}", 1).expect_err("must reject");
    assert!(err.message.contains("scalar value"), "{err}");
}

#[test]
fn very_long_input_is_handled_without_panic() {
    let long = "x".repeat(100_000);
    let line = format!("{{\"fields\":{{\"a\":\"{long}\"}},\"seq\":1,\"type\":\"A\"}}");
    let parsed = decode(&line, 1).expect("long values are legal");
    assert_eq!(parsed.fields["a"].len(), 100_000);
}
